Framework/Algorithms/src/ReflectometryReductionOneAuto3.cpp

```diff
@@ -0,0 +1,1126 @@
+      return boost::none;
+    }
+    qstep = calcRes->getProperty("Resolution");
+    qstep = -qstep;
+  }
+  return qstep;
+}
+
+/** Rebin and scale a workspace in Q.
+ *
+ * @param inputWS :: the workspace in Q
+ * @param params :: A vector containing the three rebin parameters (min, step
+ * and max)
+ * @return :: the output workspace
+ */
+MatrixWorkspace_sptr
+ReflectometryReductionOneAuto3::rebinAndScale(MatrixWorkspace_sptr inputWS,
+                                              RebinParams const &params) {
+  // Rebin
+  IAlgorithm_sptr algRebin = createChildAlgorithm("Rebin");
+  algRebin->initialize();
+  algRebin->setProperty("InputWorkspace", inputWS);
+  algRebin->setProperty("OutputWorkspace", inputWS);
+  algRebin->setProperty("Params", params.asVector());
+  algRebin->execute();
+  MatrixWorkspace_sptr IvsQ = algRebin->getProperty("OutputWorkspace");
+
+  // Scale (optional)
+  Property *scaleProp = getProperty("ScaleFactor");
+  if (!scaleProp->isDefault()) {
+    double scaleFactor = getProperty("ScaleFactor");
+    IAlgorithm_sptr algScale = createChildAlgorithm("Scale");
+    algScale->initialize();
+    algScale->setProperty("InputWorkspace", IvsQ);
+    algScale->setProperty("OutputWorkspace", IvsQ);
+    algScale->setProperty("Factor", 1.0 / scaleFactor);
+    algScale->execute();
+    IvsQ = algScale->getProperty("OutputWorkspace");
+  }
+
+  return IvsQ;
+}
+
+MatrixWorkspace_sptr
+ReflectometryReductionOneAuto3::cropQ(MatrixWorkspace_sptr inputWS,
+                                      RebinParams const &params) {
+  IAlgorithm_sptr algCrop = createChildAlgorithm("CropWorkspace");
+  algCrop->initialize();
+  algCrop->setProperty("InputWorkspace", inputWS);
+  algCrop->setProperty("OutputWorkspace", inputWS);
+  if (!params.qMinIsDefault())
+    algCrop->setProperty("XMin", params.qMin());
+  if (!params.qMaxIsDefault())
+    algCrop->setProperty("XMax", params.qMax());
+  algCrop->execute();
+  MatrixWorkspace_sptr IvsQ = algCrop->getProperty("OutputWorkspace");
+  return IvsQ;
+}
+
+/**
+ * @brief Get the Property Or return a default given value
+ *
+ * @param propertyName : the name of the property to get
+ * @param defaultValue : the default value to use if the property is not set
+ * @param isDefault [out] : true if the default value was used
+ */
+double ReflectometryReductionOneAuto3::getPropertyOrDefault(
+    const std::string &propertyName, const double defaultValue,
+    bool &isDefault) {
+  Property *property = getProperty(propertyName);
+  isDefault = property->isDefault();
+  if (isDefault)
+    return defaultValue;
+  else
+    return getProperty(propertyName);
+}
+
+/** Check if input workspace is a group
+ */
+bool ReflectometryReductionOneAuto3::checkGroups() {
+
+  const std::string wsName = getPropertyValue("InputWorkspace");
+
+  try {
+    auto ws =
+        AnalysisDataService::Instance().retrieveWS<WorkspaceGroup>(wsName);
+    if (ws)
+      return true;
+  } catch (...) {
```

Comment:
Do we need to be so vague when catching exceptions here?

Would it be better to check if a workspace with the given name exists first?

Reply:
Similarly with this one